doc/tests-updates.md

```diff
@@ -0,0 +1,381 @@
+loads.
+
+- Easy new test addition.
+
+- Easy test case failure analysis.
+
+- Generate a few tests based on singe test description. E.g. in fussing test
+it's required to know, how the Tempesta would react on a message, and what
+should be threated as error. The 'splitting on different skbs' is actually
+fussing test feature, not functional. Usually it also make sense to generate
+workload test based on functional profile.
+
+
+## Problems of current implementation
+
+The main problem is having two frameworks in the same time. Old, feature rich,
+but too sophisticated, with requirement to understand 5-6 files to understand
+what the test is actually does. And a new, more friendly, but not complete and
+with lack of the most important features. Transition of an old one to a new
+wasn't completed, and the development of both was rather unorganized, thus
+new framework is becoming more and more complex and loses its advantages.
+
+In both frameworks every test type is implemented as unique test, so it's
+required to write message integrity, fussing, workload and many other test
+cases for the same situation.
+
+Heavy dependency between tests. When a new feature is added, it's added to the
+test framework core and affects a lot of tests. Sometimes unpredictably and
+badly.
+
+
+## Proposed design
+
+The proposed design improves the new test framework and intends to deliver it to
+full-feature test tool.
+
+The whole test case is described in single file, configuration of each entity
+is listed as plain text template with some parameters. No procedure generated
+configs like in old framework.
+
+Most of the test cases should use following description:
+```
+
+class SomeTest(BaseTestClass):
+
+    backends = [# For all tests except stress tests:
+                {
+                    'type'      : 'deproxy',
+                    'id'        : 'id1',
+                    'response'  : gen_response_func,
+                },
+                # OR:
+                {
+                    'type'      : 'wrk',
+                    'id'        : 'id1',
+                    'script'    : script_name,
+                    # Other wrk-specific options.
+                },
+               ]
+    clients  = [# For all tests except stress tests:
+                {
+                    'type'      : 'deproxy',
+                    'id'        : 'id1',
+                    'request'   : gen_request_func,
+                    'response'  : response_cb,
+                },
+                # OR:
+                {
+                    'type'      : 'wrk'
+                    'id'        : 'id1',
+                    'script'    : script_name,
+                    # Other wrk-specific options.
+                },
+               ]
+    tempesta = {
+                'config' : '...'
+               }
+    test_variations = ['functional', 'workload', 'fussing']
+
+    def test(self):
+        < ... test procedure steps ... >
+        pass
+```
```

Comment:
`gen_response_func` and `gen_request_func` are nice and must be quite useful, but it seems they also can be done as an extension (feature) of current framework.

Reply:
**Update** It seems the callbacks are only needed to generate different requests and responses for single deproxy/tempesta run. Actually `test_malformed_headers.py` and `test_tls_integrity.py` use current API for the same things: we have `set_response()` to dynamically change deproxy response also each response is trivially checked.

Please review `test_tls_integrity` in #103 and let's discuss there if the architecture can benfit from the callbacks.

Reply:
The `set_response()` callback in the new framework much near to the behaviour I want. But it still has a lot of things to enhance (as I think). Tests in the current framework are not scalable, you cant run the same test on one or 1'000 connections simultaneously.
 
How the current tests in the new framework run:
- You start all servers/clients;
- generate requests and responses and assign them to client/server (callback `set_response()` you've mentioned above),
- press `Run` button: wait until all server connections are established and begin to start requests, wait until all messages are sent;
- run asserts, on message integrity and message content, e.g. md5 sum checks against received body in `test_tls_integrity` test. The tests contain no headers validation by default and it must be implemented. 

How to make test scalable:
- configure a number of callback functions for requests and responses;
- start servers and clients, each in it's own process.
- press `Run` button:
    - client makes a callback to generate request (or to chose one of predefined one) and send it to server;
    - server receives request and runs the callback to validate it, compare with expected, generate response for it. Server sends  the response;
    - client calls a callback function to validate received response and compare it with the expected;
    - Probably callbacks on connection closes are also required.
- Just finish the tests: all assertions are done in client or server callbacks. 

The second case is more scalable, the test itself is tolerant to number of  concurrent connections. May be it's a my misbelief, but I think that scalability from one connection to thousands is important. Issue #107 is very tightly connected here.  It's a huge work while we have a plenty of other tasks, I understand that. But I don't know more suitable tools here.

Sure, some special changes will be required to make tests scalable, e.g. in cache tests each client must request unique URIs; in scheduler tests we would like to see backend server ID in the response, anyway it's better than check request counters across all backends.